src/areastore.cpp

```diff
@@ -0,0 +1,328 @@
+}
+
+u32 AreaStore::getFreeId(v3s16 minedge, v3s16 maxedge)
+{
+	int keep_on = 100;
+	while (keep_on--) {
+		m_highest_id++;
+		// Handle overflows, we dont want to return 0
+		if (m_highest_id == AREA_ID_INVALID)
+			m_highest_id++;
+		if (areas_map.find(m_highest_id) == areas_map.end())
+			return m_highest_id;
+	}
+	// search failed
+	return AREA_ID_INVALID;
+}
+
+const Area *AreaStore::getArea(u32 id) const
+{
+	const Area *res = NULL;
+	std::map<u32, Area>::const_iterator itr = areas_map.find(id);
+	if (itr != areas_map.end()) {
+		res = &itr->second;
+	}
+	return res;
+}
+
+#if 0
+bool AreaStore::deserialize(std::istream &is)
+{
+	u8 ver = readU8(is);
+	if (ver != 1)
+		return false;
+	u16 count_areas = readU16(is);
+	for (u16 i = 0; i < count_areas; i++) {
+		// deserialize an area
+		Area a;
+		a.id = readU32(is);
+		a.minedge = readV3S16(is);
+		a.maxedge = readV3S16(is);
+		a.datalen = readU16(is);
+		a.data = new char[a.datalen];
+		is.read((char *) a.data, a.datalen);
+		insertArea(a);
+	}
+	return true;
+}
+
+
+static bool serialize_area(void *ostr, Area *a)
+{
+	std::ostream &os = *((std::ostream *) ostr);
+	writeU32(os, a->id);
+	writeV3S16(os, a->minedge);
+	writeV3S16(os, a->maxedge);
+	writeU16(os, a->datalen);
+	os.write(a->data, a->datalen);
+
+	return false;
+}
+
+
+void AreaStore::serialize(std::ostream &os) const
+{
+	// write initial data
+	writeU8(os, 1); // serialisation version
+	writeU16(os, areas_map.size()); //DANGER: not platform independent
+	forEach(&serialize_area, &os);
+}
+
+#endif
+
+void AreaStore::invalidateCache()
+{
+	if (cache_enabled) {
+		m_res_cache.invalidate();
+	}
+}
+
+void AreaStore::setCacheParams(bool enabled, u8 block_radius, size_t limit)
+{
+	cache_enabled = enabled;
+	m_cacheblock_radius = MYMAX(block_radius, 16);
+	m_res_cache.setLimit(MYMAX(limit, 20));
+	invalidateCache();
+}
+
+void AreaStore::cacheMiss(void *data, const v3s16 &mpos, std::vector<Area *> *dest)
+{
+	AreaStore *as = (AreaStore *)data;
+	u8 r = as->m_cacheblock_radius;
+
+	// get the points at the edges of the mapblock
+	v3s16 minedge(mpos.X * r, mpos.Y * r, mpos.Z * r);
+	v3s16 maxedge(
+		minedge.X + r - 1,
+		minedge.Y + r - 1,
+		minedge.Z + r - 1);
+
+	as->getAreasInArea(dest, minedge, maxedge, true);
+
+	/* infostream << "Cache miss with " << dest->size() << " areas, between ("
+			<< minedge.X << ", " << minedge.Y << ", " << minedge.Z
+			<< ") and ("
+			<< maxedge.X << ", " << maxedge.Y << ", " << maxedge.Z
+			<< ")" << std::endl; // */
+}
+
+void AreaStore::getAreasForPos(std::vector<Area *> *result, v3s16 pos)
+{
+	if (cache_enabled) {
+		v3s16 mblock = getContainerPos(pos, m_cacheblock_radius);
+		const std::vector<Area *> *pre_list = m_res_cache.lookupCache(mblock);
+
+		size_t s_p_l = pre_list->size();
+		for (size_t i = 0; i < s_p_l; i++) {
+			Area *b = (*pre_list)[i];
+			if (AST_CONTAINS_PT(b, pos)) {
+				result->push_back(b);
+			}
+		}
+	} else {
+		return getAreasForPosImpl(result, pos);
+	}
+}
+
+
```

Comment:
Hmm, maybe there should be a bit more visual separation between AreaStore interface functions and implementations, if you prefer.  I would personally put a comment here that says in effect "this is VectorAreaStore related stuff"
```
////
//// VectorAreaStore
////

/*
 * VectorAreaStore
 */
```
etc.